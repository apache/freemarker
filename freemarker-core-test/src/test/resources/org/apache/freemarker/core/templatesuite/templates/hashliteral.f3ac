<#--
  Licensed to the Apache Software Foundation (ASF) under one
  or more contributor license agreements.  See the NOTICE file
  distributed with this work for additional information
  regarding copyright ownership.  The ASF licenses this file
  to you under the Apache License, Version 2.0 (the
  "License"); you may not use this file except in compliance
  with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing,
  software distributed under the License is distributed on an
  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  KIND, either express or implied.  See the License for the
  specific language governing permissions and limitations
  under the License.
-->
<html>
<head>
<title>FreeMarker: Hash Literal Test</title>
</head>
<body>

<p>A simple test follows:</p>

<p>${message}</p>

<p>Now perform a hash assignment:</p>

<#assign mymessage = "hello", foo="bar", one="1">
<#assign test = { "test1": "test23", 
                 "test45" : message, 
                  mymessage : "hello all", 
                  foo: one}>

${test.test1}
${test.test45}
${test.hello}
${test.bar}

<p>Now update the assignment and repeat:</p>

<#assign mymessage = "world">

${test.test1}
${test.test45}
${test.hello}

${test.bar}

<p>Now reassign the list and repeat:</p>

<#assign hash= {"temp" : "Temporary"}>
<#assign test = { "test1" : "test23", 
                        "test45" : message, 
                        mymessage : "hello all", 
                        foo : one, 
                        "hash" : hash[ "temp" ], 
                        "true" : hash.temp, 
                        "newhash" : hash}>

${test.test1}
${test.test45}
${test.hello!}

${test.bar}
${test.hash}
${test.true}
${test.newhash.temp}

<p>Pathological case: zero item hash:</p>

<#assign test = {}>
${test.test1!}

<p>Hash of number literals:</p>
<#assign test = {"1" : 2}>
${test["1"]}

<p>Hash concatenation:</p>
<#assign cc = { "a" : 1, "b" : 2 } + { "b" : 3, "c" : 4 }>
<#list cc as key, value>
${key} => ${value}
</#list>

<p>Empty hash concatenation:</p>
${({} + { "a" : "foo" }).a}, ${({ "a" : "bar" } + {}).a}

</body>
</html>
<@noOutput>

<#assign m = { 'a': 1, 'b', 2, 'a': 3 }>
<@assertEquals expected="a, b" actual=m?keys?join(', ') />
<@assertEquals expected="3, 2" actual=m?values?join(', ') />
<@assertEquals expected=3 actual=m['a'] />
<@assertEquals expected=2 actual=m['b'] />

</@noOutput>