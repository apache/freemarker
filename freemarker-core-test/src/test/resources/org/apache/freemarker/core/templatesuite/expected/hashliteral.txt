/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
<html>
<head>
<title>FreeMarker: Hash Literal Test</title>
</head>
<body>

<p>A simple test follows:</p>

<p>Hello, world!</p>

<p>Now perform a hash assignment:</p>


test23
Hello, world!
hello all
1

<p>Now update the assignment and repeat:</p>


test23
Hello, world!
hello all

1

<p>Now reassign the list and repeat:</p>


test23
Hello, world!


1
Temporary
Temporary
Temporary

<p>Pathological case: zero item hash:</p>



<p>Hash of number literals:</p>
2

<p>Hash concatenation:</p>
a => 1
b => 3
c => 4

<p>Empty hash concatenation:</p>
foo, bar

</body>
</html>
